### PR TITLE
Dialog2 no longer accepts styled system props

### PR DESCRIPTION
**Pull request by jfuchs**
This PR updates Dialog2 (AKA `src/Dialog/Dialog.tsx`) to no longer accept system props.

See github/primer#296

### Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

File: docs/content/Dialog2.mdx [hunk 1]
```diff
@@ -87,6 +87,7 @@ By default, the Dialog component implements the design and interactions defined
...
 | width         | `"small" │ "medium" │ "large" │ "xlarge"`      | `"xlarge"` | The fixed width of the dialog.                                                                                                                                                                                                                                          |
 | height        | `"small" │ "large" │ "auto"`                   | `"auto"`   | The fixed height of the dialog, or, auto to adjust the height based on its contents.                                                                                                                                                                                    |
+| sx            | SystemStyleObject                              | {}         | Style to be applied to the component                                                                                                                                                                                                                                    |
```
Comment (mperrotti) [inline docs/content/Dialog2.mdx:90]: Non-blocking nitpick: I would have put "SystemStyleObject" to show it as preformatted text.

File: src/__tests__/Dialog2.types.test.tsx [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import React from 'react'
+import {Dialog} from '../Dialog/Dialog'
+
+export function shouldAcceptCallWithNoProps() {
+  return <Dialog onClose={() => null} />
+}
+
+export function shouldNotAcceptSystemProps() {
+  // @ts-expect-error system props should not be accepted
```
Comment (mperrotti) [inline src/__tests__/Dialog2.types.test.tsx:9]: Wow, I didn't know `@ts-expect-error` was a thing. Cool!

**Review by mperrotti — APPROVED** [review]

File: docs/content/Dialog2.mdx (hunk 1 above)
Comment (jfuchs) [reply docs/content/Dialog2.mdx:90]: Ohh that's a good call. I was just chatting with @colebemis the other day about how we could make some react components to build prop tables in a semi-automated way.
```suggestion
| sx            | `SystemStyleObject`                            | {}         | Style to be applied to the component                                                                                                                                                                                                                                    |
```

File: src/__tests__/Dialog2.types.test.tsx (hunk 2 above)
Comment (jfuchs) [reply src/__tests__/Dialog2.types.test.tsx:9]: I love writing type tests now! Given how many regressions I've run into around types, it's nice to have some more confidence!

**Pull request merged**